dev/tools/bin/find_commit.dart

```diff
@@ -0,0 +1,102 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// This script looks at the current commit and branch of the git repository in
+// which it was run, and finds the contemporary commit in the master branch of
+// another git repository, whose path is provided on the command line. The
+// contemporary commit is the one that was public at the time of the last commit
+// on the master branch before the current commit's branch was created.
+
+import 'dart:io';
+
+const bool debugLogging = false;
+
+void log(String message) {
+  if (debugLogging)
+    print(message);
+}
+
+class Commit {
+  Commit(this.hash, this.timestamp);
+
+  final String hash;
+  final DateTime timestamp;
+
+  static String formatArgument = '--format=%H %cI';
+
+  static Commit parse(String line) {
+    final int space = line.indexOf(' ');
+    return Commit(line.substring(0, space), DateTime.parse(line.substring(space+1, line.length).trimRight()));
+  }
+
+  static List<Commit> parseList(String lines) {
+    return lines.split('\n').where((String line) => line.isNotEmpty).map(parse).toList().reversed.toList();
+  }
+}
+
+String findCommit({
+  required String primaryRepoDirectory,
+  required String primaryBranch,
+  required String primaryTrunk,
+  required String secondaryRepoDirectory,
+  required String secondaryBranch,
+}) {
+  final Commit anchor;
```

Comment:
This passes analysis both on CI and locally for me, but how are you able to declare this `final` and then assign it a value later?

Reply:
> this is a new feature with null safety! If the compiler can prove that it'll be initialized before it's used, you can declare it final without an initializer. (It's like `late` but the compiler can prove it's safe, whereas `late` is like `dynamic` in that it tells the compiler "trust me I know what I'm doing" and therefore is probably a source of bugs.)

mind blown.